README.md
# VSCode FileHeader

[![Markeetplace Badge][marketplace-badge]][marketplace] [![Install][install-badge]][marketplace] [![Download][download-badge]][marketplace] ![Size][size-badge] [![BSD License Badge][license-badge]][license]

[中文文档](./README-zh_CN.md) | [英文文档](./README.md)

This is a header extension that contains multiple languages.

Feel good friends please give a star ~

* Usage [cookiecutter-vscode](https://github.com/caizhengxin/cookiecutter-vscode) create project.
* [Custom fileheader Extension template](https://github.com/caizhengxin/fileheader-template)

## Install

```bash
1. ctrl + shift + x
2. Search VSCodeFileHeader
```

or

```bash
1. ctrl + p
2. ext install jankincai.vscodefileheader
```

## Features

* Support for automatic update time and author
* Support for custom template(header/body)
* Support for multiple languages
* Support for suffix name mapping templates
* Support for ignore suffix
* Support for manual and automatic insert comment
* [Support vscode variables](https://code.visualstudio.com/docs/editor/variables-reference)
* Support remote synchronization(install git)

## Example

![example](./images/example.gif)

![example](./images/custom-example.gif)

[More examples](./demo)

## Language

* ActionScript
* AppleScript
* ASP
* Batch
* C
* C#
* C++
* Clojire
* CSS
* D
* Dart
* Erlang
* Go
* Haskell
* HTML
* Java
* JavaScript
* LaTeX
* Lisp
* Lua
* Matlab
* Objective-C
* OCaml
* Pascal
* Perl
* PHP
* Python
* R
* Ruby
* Scala
* SCSS
* ShellScript
* SQL
* TypeScript
* Vue
* XML
* YAML

## Usage

```python
# Settings
{
    "fileheader.author": "JanKinCai",
}

# Save file or ``ctrl + alt + i`` insert comment
```

## Extension Settings

```python
{
    # Set author
    "fileheader.author": "Your name",

    # Set insert body switch, default false
    "fileheader.body": true,

    # Set open file to insert header comment, default false
    "fileheader.open": true,

    # Set save file to insert header comment, default true
    "fileheader.save": true,

    # find max line of header comment, default 10
    "fileheader.header_max_line": 10,

    # Set datetime format, default YYYY-MM-DD HH:mm:ss
    # moment.js
    "fileheader.dateformat": "YYYY-MM-DD HH:mm:ss",

    # Set template mapping
    "fileheader.file_suffix_mapping": {
        ".pyx": "Python",
        "jkc-*.h": "H"
    },

    "fileheader.ignore": [
        "*.txt",             # Ignore suffix name
        "test.py",           # Ignore file
        "test"               # Ignore dir
    ]
}

```

### Custom template(Add Email)

Note: ``The template header connot be modified and can be added.``

Custom Header template ``template/header/Python.tmpl``:

```conf
# @Author: {{author}}
# @Date:   {{create_time}}
# @Email:  {{email}}
# @Last Modified by:   {{last_modified_by}}
# @Last Modified time: {{last_modified_time}}
```

Custom Body template ``template/body/Python.tmpl``:

```conf


def main():
    print("Hello, World!")


if __name__ == "__main__":
    main()
```

Settings:

```python
{
    # Set template path
    "fileheader.custom_template_path": "xxx/template/",
    # Set other config
    "fileheader.other_config": {
        "email": "Your email",
    }
    # Set suffix mapping
    "fileheader.file_suffix_mapping": {
        ".py": "Python"
    },
}
```

* [Custom fileheader Extension template](https://github.com/caizhengxin/fileheader-template)

### fully modify template (example)

Custom Header template template/header/JKC.tmpl:

```conf
# @Name: {{name}}
```

Custom Header template template/body/JKC.tmpl:

```conf
```

settings:

```python
{
    "fileheader.is_header_exists": "@Name:",
    "fileheader.other_config": {
        "name": "jankincai",
    },
    "fileheader.file_suffix_mapping": {
        ".jkc": "JKC"
    },   
}
```

### vscode variables

https://code.visualstudio.com/docs/editor/variables-reference

Custom Header template ``template/header/Python.tmpl``:

```conf
# @Author: {{author}}
# @Date:   {{create_time}}
# @Last Modified by:   {{last_modified_by}}
# @Last Modified time: {{last_modified_time}}
# {{workspaceFolder}}
# {{workspaceFolderBasename}}
# {{file}}
# {{relativeFile}}
# {{relativeFileDirname}}
# {{fileBasename}}
# {{fileBasenameNoExtension}}
# {{fileDirname}}
# {{fileExtname}}
# {{cwd}}
```

### Sync remote template

```python
{
    # Set template path
    "fileheader.custom_template_path": "xxx/template/",

    # set ssh, https://jdblischak.github.io/2014-09-18-chicago/novice/git/05-sshkeys.html
    # "fileheader.remote": "git@github.com:caizhengxin/vscodefileheader.git"

    "fileheader.remote": "https://github.com/caizhengxin/vscodefileheader.git"
}
```

[marketplace]: https://marketplace.visualstudio.com/items?itemName=jankincai.vscodefileheader#review-details
[marketplace-badge]: https://vsmarketplacebadge.apphb.com/version-short/jankincai.vscodefileheader.svg?style=flat-square
[install-badge]: https://img.shields.io/visual-studio-marketplace/i/jankincai.vscodefileheader?style=flat-square
[download-badge]: https://img.shields.io/visual-studio-marketplace/d/jankincai.vscodefileheader?style=flat-square
[size-badge]: https://img.shields.io/github/languages/code-size/caizhengxin/vscodefileheader?style=flat-square
[license]: ./LICENSE
[license-badge]: https://img.shields.io/github/license/caizhengxin/vscodefileheader?style=flat-square


CHANGELOG.md
# Change Log

All notable changes to the "vscodefileheader" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

## [0.9.1] - 2021-07-06

### Fixed
- Fix insert template bug

## [0.9.0] - 2021-04-25
### Added

- Add create/open template

### Fixed

- Fix template bug
- Fix ignore bug

### Changed

- Refactor code structure

## [0.8.0] - 2021-03-18
### Added

- Add case conversion support
- Add template function replace
- Add C language header file template
- Add support for wildcard * template mapping
- Add template inheritance

## [0.7.0] - 2021-03-11
### Added

- Add ``Dart`` language template

### Changed

- Update header only file has been modified
- Use file creation date as the ``create_time``
- Update template style

## [0.6.1] - 2020-11-28
### Fixed

- Fix custom template bug(#9)

## [0.6.0] - 2020-11-17
### Added

- Support remote synchronization
- Add keywords ``header_max_line`` and `is_header_exists`` to the configuration file

## [0.5.1] - 2020-09-28
### Fixed

- ignore file, if the file header comment exists, it will be updated

## [0.5.0] - 2020-06-28
### Added

- [Support vscode variables](https://code.visualstudio.com/docs/editor/variables-reference)

## [0.4.1] - 2020-03-14
### Fixed

- Support vscode 1.43.0

## [0.4.0] - 2020-01-03
### Added

- If exists ``# -*- coding: utf-8 -*-``, first delete and insert python header comment.
- Add rust template
- Add php tail tag, body is not opened

### Changed

- Update template format
- Update ignore achieve
- Update function comment

## [0.3.2] - 2019-12-26
### Fixed

- When the editor does not have enough lines, find the line number error

## [0.3.1] - 2019-12-26
### Fixed

- Fix php tag bug (#1)

## [0.3.0] - 2019-12-6
### Added

- Add custom template header field mapping

## [0.2.7] - 2019-12-5
### Fixed

- Update php-template

## [0.2.6] - 2019-11-20
### Fixed

- First match file name and suffix.

## [0.2.0] - 2019-11-4
### Added

- Add ``TypeScript`` comment
- Add ``Vue`` comment
- Add manual and automatic insert comment

## [0.1.0] - 2019-8-31
### Added

- Add automatic update time and author
- Add custom template(header/body)
- Add suffix name mapping templates
- Add ignore suffix

[unreleased]: https://github.com/caizhengxin/vscodefileheader/compare/v0.9.1...HEAD
[0.9.1]: https://github.com/caizhengxin/vscodefileheader/compare/v0.9.0...v0.9.1
[0.9.0]: https://github.com/caizhengxin/vscodefileheader/compare/v0.8.0...v0.9.0
[0.8.0]: https://github.com/caizhengxin/vscodefileheader/compare/v0.7.0...v0.8.0
[0.7.0]: https://github.com/caizhengxin/vscodefileheader/compare/v0.6.1...v0.7.0
[0.6.1]: https://github.com/caizhengxin/vscodefileheader/compare/v0.6.0...v0.6.1
[0.6.0]: https://github.com/caizhengxin/vscodefileheader/compare/v0.5.1...v0.6.0
[0.5.1]: https://github.com/caizhengxin/vscodefileheader/compare/v0.5.0...v0.5.1
[0.5.0]: https://github.com/caizhengxin/vscodefileheader/compare/v0.4.1...v0.5.0
[0.4.1]: https://github.com/caizhengxin/vscodefileheader/compare/v0.4.0...v0.4.1
[0.4.0]: https://github.com/caizhengxin/vscodefileheader/compare/v0.3.2...v0.4.0
[0.3.2]: https://github.com/caizhengxin/vscodefileheader/compare/v0.3.1...v0.3.2
[0.3.1]: https://github.com/caizhengxin/vscodefileheader/compare/v0.2.7...v0.3.1
[0.2.7]: https://github.com/caizhengxin/vscodefileheader/compare/v0.2.6...v0.2.7
[0.2.6]: https://github.com/caizhengxin/vscodefileheader/compare/v0.2.3...v0.2.6
[0.2.0]: https://github.com/caizhengxin/vscodefileheader/compare/v0.1.8...v0.2.1
[0.1.0]: https://github.com/caizhengxin/vscodefileheader/compare/v0.0.2...v0.1.5


demo/test.c
/*
* @Author: JanKinCai
* @Date:   2019-11-11 23:45:17
* @Last Modified by:   JanKinCai
* @Last Modified time: 2019-11-11 23:45:20
*/
#include <stdio.h>


int main(int argc, char const *argv[])
{
    /* code */
    return 0;
}



demo/test.r
# @Author: JanKinCai
# @Date:   2019-12-31 17:16:58
# @Last Modified by:   JanKinCai
# @Last Modified time: 2019-12-31 17:16:58



demo/test.cs
/*
* @Author: JanKinCai
* @Date:   2020-01-03 22:21:43
* @Last Modified by:   JanKinCai
* @Last Modified time: 2020-01-03 22:21:43
*/


class MainClass {
    static void Main() {

    }
}



src/editor.ts
/**
 * @Author: JanKinCai
 * @Date:   2021-04-24 16:57:18
 * @Last Modified by:   JanKinCai
 * @Last Modified time: 2021-04-25 01:40:20
 */
import * as path from 'path';
import * as vscode from 'vscode';
import * as moment from 'moment';
import * as fs from 'fs';


/*
    * root
    * dir
    * base
    * ext
    * name
    */


class editorObject {
    editor: vscode.TextEditor;
    config: vscode.WorkspaceConfiguration;
    pathobj: any;
    datetimeFormat: string;

    constructor(editor: any, config: any) {
        this.editor = editor;
        this.pathobj = undefined;

        if (this.editor) {
            this.pathobj = path.parse(this.editor.document.fileName);
        }
        this.config = config || vscode.workspace.getConfiguration("fileheader");
        this.datetimeFormat = config.dateformat
    }

    isEditor(): boolean {
        return this.editor === undefined;
    }

    isSuffix(s: any): boolean {
        return this.pathobj.ext.lastIndexOf(s) !== -1;
    }

    isSuffixList(s: string[]): boolean {
        return s.indexOf(this.pathobj.ext) !== -1;
    }

    isIgnore(s: string[]): boolean {
        for(let ige of s){
            let reg: any = new RegExp(ige.replace(".", "\\.").replace("*", ".*"));
    
            if(reg.test(this.pathobj.base) || reg.test(path.join(this.pathobj.dir, this.pathobj.base))){
                return true;
            }
        }
    
        return false;    
    }

    getDateTime(): string {
        return moment().format(this.datetimeFormat);
    }

    getFileBirthDateTime(): string {
        let fileStat: any = fs.statSync(this.editor.document.fileName);
        let birthTime: string = fileStat.birthtime;
    
        if (birthTime.toString().startsWith("1970")) {
            birthTime = fileStat.ctime; // When birthtime is not available
        }
    
        return moment(birthTime).format(this.datetimeFormat);    
    }

    getSuffix(): string {
        return this.pathobj.ext.toLowerCase() || this.pathobj.name.toLowerCase();    
    }

    getFileName(): string {
        return this.pathobj.name.toLowerCase();
    }

    getMaxHeaderLine(): number {
        return Math.min(this.editor.document.lineCount, this.config.header_max_line);
    }

    findStringLine(text: string, max_line: number = this.config.header_max_line): number {
        let lineCount: number =  Math.min(max_line, this.editor.document.lineCount);
        let i: number = 0;
    
        for(i = 0; i <= lineCount - 1; i++){
            if (this.editor.document.lineAt(i).text.toLowerCase().indexOf(text.toLowerCase()) !== -1){
                return i;
            }
        }
    
        return -1;    
    }

    deleteEditorComment(text: string, max_line: number): void {
        let line: number = this.findStringLine(text, max_line);

        if(line !== -1){
            this.editor.edit((editobj: any) => {
                editobj.delete(new vscode.Range(line, 0, line, text.length));
            });
        }
    }

    insertEditorComment(text: string): void {
        let line: number = this.editor.document.lineCount + 1;
        this.editor.edit((editobj: any) => {
            editobj.delete(new vscode.Range(line, 0, line, text.length));
        });    
    }

    isHeaderExists(): boolean {
        if (this.config.is_header_exists && this.findStringLine(this.config.is_header_exists, this.getMaxHeaderLine()) !== -1) {
            return true;
        }
    
        if(this.findStringLine("@Author:", this.getMaxHeaderLine()) !== -1){
            return true;
        }
    
        return false;    
    }
}


export {
    editorObject
}

src/command.ts
/**
 * @Author: JanKinCai
 * @Date:   2021-04-24 16:24:09
 * @Last Modified by:   JanKinCai
 * @Last Modified time: 2021-04-25 01:31:07
 */
import * as child_process from 'child_process';
import * as vscode from 'vscode';
import * as path from 'path';
import * as fs from 'fs';
import * as template from './template';


function createTemplateCommand(config: any, file_suffix_mapping: any): void {
    let editor: any = vscode.window.activeTextEditor;
    let tmplobj = new template.Template(editor, config, file_suffix_mapping);

    if (config.custom_template_path) {
        vscode.window.showInputBox({ignoreFocusOut: true, password: false, prompt: "Please type template name"}).then(tmplname => {
            if (tmplname === undefined || tmplname.trim() === '') {
                vscode.window.showInformationMessage("Please type template name")
            }
            else {
                tmplobj.create(tmplname);
            }
        });	
    }
    else {
        vscode.window.showInformationMessage("Please set vscodefileheader custome template path.");
    }
}


function openTemplateCommand(config: any, file_suffix_mapping: any): void {
    let editor: any = vscode.window.activeTextEditor;

    if (editor) {
        let tmplobj = new template.Template(editor, config, file_suffix_mapping);

        tmplobj.open();
    }
    else {
        vscode.window.showInformationMessage("No active window");
    }
}


/**
 * Sync template Command
 * 
 * @return void
 */
 function syncTemplateCommand(config: any, file_suffix_mapping: any): void {

	if (config.custom_template_path && config.remote) {

		if (!fs.existsSync(config.custom_template_path)) {
			fs.mkdirSync(config.custom_template_path, {recursive: true});

			/* git clone */
			child_process.exec(`git clone ${config.remote} ${config.custom_template_path}`);
		}
		else {
			child_process.exec(`cd ${config.custom_template_path}`);
			child_process.exec(`git pull origin master`);
		}

		/* Read file_suffix_map.json or file_suffix_mapping.json */
		let file_suffix_mapping_path: string = path.join(config.custom_template_path, "file_suffix_map.json");

		if (!fs.existsSync(file_suffix_mapping_path))
		{
            file_suffix_mapping_path = path.join(config.custom_template_path, "file_suffix_mapping.json");
		}

        if (fs.existsSync(file_suffix_mapping_path))
        {
            Object.assign(file_suffix_mapping, require(file_suffix_mapping_path));
        }
	}
}


function updateTemplateCommand(config: any, file_suffix_mapping: any): void {
    let editor: any = vscode.window.activeTextEditor;

    if (editor) {
        let tmplobj = new template.Template(editor, config, file_suffix_mapping);

        tmplobj.update();
    }
    else {
        vscode.window.showInformationMessage("No active window");
    }
}


function updateTemplateCommand2(config: any, file_suffix_mapping: any): void {
    let editor: any = vscode.window.activeTextEditor;

    if (editor) {
        let tmplobj = new template.Template(editor, config, file_suffix_mapping);

        tmplobj.update2();
    }
}


function insertTemplateCommand(config: any, file_suffix_mapping: any): void {
    let editor: any = vscode.window.activeTextEditor;

    let tmplobj = new template.Template(editor, config, file_suffix_mapping);

    tmplobj.insert();

    if (editor) {
        let tmplobj = new template.Template(editor, config, file_suffix_mapping);

        tmplobj.insert();
    }
    else {
        vscode.window.showInformationMessage("No active window");
    }
}


export {
    createTemplateCommand,
    openTemplateCommand,
    syncTemplateCommand,
    updateTemplateCommand,
    insertTemplateCommand,
    updateTemplateCommand2,
}


src/template.ts
/**
 * @Author: JanKinCai
 * @Date:   2021-04-22 23:41:46
 * @Last Modified by:   JanKinCai
 * @Last Modified time: 2021-07-06 13:39:42
 */
import * as vscode from 'vscode';
import * as path from 'path';
import * as fs from 'fs';
import * as art_template from 'art-template';
import * as editor from './editor';


class Template extends editor.editorObject {
    default_tmpl_path: string;
    file_suffix_mapping: any;

    constructor(editor: any, config: any, file_suffix_mapping: any) {
        super(editor, config);
        this.default_tmpl_path = this.getDefaultTemplatePath();
        this.file_suffix_mapping = file_suffix_mapping;
    }

    /**
     * getSuffix
     * 
     * @return string 
     */
    getSuffix(): string {
        let pathobj:any = path.parse(this.editor.document.fileName);

        return pathobj.ext.toLowerCase() || pathobj.name.toLowerCase();    
    }

    /**
     * getFileName
     * 
     * @return string 
     */
    getFileName(): string {
        return path.parse(this.editor.document.fileName).name.toLowerCase();
    }

    /**
     * getDefaultTemplatePath
     * 
     * @return string 
     */
    getDefaultTemplatePath(): string {
        return path.join(path.dirname(__dirname), "template");
    }

    getDefaultArtTemplate(type: string="header"): string {
        return path.join(this.default_tmpl_path, "art", type + ".art");
    }

    /**
     * getTemplatePath
     * 
     * @param type (string): header/body
     * 
     * @return string | undefined
     */
    getTemplatePath(type: string="header"): string | undefined {
        let suffix: string = this.getSuffix();
        let name: string = this.getFileName();
        let tmpl: string = "";
    
        for (let v in this.config.file_suffix_mapping) {
            let reg: any = new RegExp("^" + v.replace(".", "\\.").replace("*", ".*") + "$");
    
            if (reg.test(name + suffix) || reg.test(suffix)) {
                tmpl = this.config.file_suffix_mapping[v] + ".tmpl"
                break;
            }
        }

        if (!tmpl) {
            tmpl = (this.file_suffix_mapping[name + suffix] || this.file_suffix_mapping[suffix]) + ".tmpl";
        }    

        if (tmpl) {
            let tmplpath = path.join(this.config.custom_template_path || "", type, tmpl);

            if (fs.existsSync(tmplpath)) {
                return tmplpath;
            }

            tmplpath = path.join(this.default_tmpl_path, type, tmpl);

            if (fs.existsSync(tmplpath)) {
                return tmplpath;
            }
        }

        return undefined;
    }

    private _create(name: string, type: string="header"): void {
		let srcpath = vscode.Uri.file(this.getDefaultArtTemplate(type));
		let dstpath = vscode.Uri.file(path.join(this.config.custom_template_path, type, name + ".tmpl"));
		vscode.workspace.fs.copy(srcpath, dstpath);
    }

    private _open(name: string, type: string="header"): void {
        let tmplpath = path.join(this.config.custom_template_path, type, name + ".tmpl");

        let textDocument: any = vscode.workspace.openTextDocument(vscode.Uri.file(tmplpath));

        let viewcolumn: number = vscode.ViewColumn.One;

        if (type === "body") {
            viewcolumn = vscode.ViewColumn.Two;
        }

        vscode.window.showTextDocument(textDocument, viewcolumn);
    }

    private _open2(tmplpath: string | undefined, type: string="header"): void {
        if (tmplpath === undefined || !fs.existsSync(tmplpath)) {
            vscode.window.showInformationMessage("Template path does not exist");
        }
        else {
            let textDocument: any = vscode.workspace.openTextDocument(vscode.Uri.file(tmplpath));

            let viewcolumn: number = vscode.ViewColumn.One;

            if (type === "body") {
                viewcolumn = vscode.ViewColumn.Two;
            }

            vscode.window.showTextDocument(textDocument, viewcolumn);
        }
    }

    create(name: string): void {
        this._create(name, "header");
        this._create(name, "body");
        this._open(name, "header");
        this._open(name, "body");
    }

    open(): void {
        let tmplpath = this.getTemplatePath("header");

        this._open2(tmplpath, "header");

        tmplpath = this.getTemplatePath("body");

        this._open2(tmplpath, "body");
    }

    private deleteEditorComments(): void {
        if(this.isSuffix(".php")){
            this.deleteEditorComment("<?php", 2);
        }else if(this.isSuffixList([".py", ".pxd", ".pyx"])){
            this.deleteEditorComment("# -*- coding: utf-8 -*-", 1);
        }    
    }

    private insertEndComments(): void {
        let lineCount: number = this.editor.document.lineCount;

        if(this.config.body && lineCount <= 1){
            if(this.isSuffix(".php")){
                this.insertEditorComment("?>");
            }
        }    
    }

    /**
     * Support Predefined variables: https://code.visualstudio.com/docs/editor/variables-reference
     */
    private predefinedVariables(): any {
        const rfd: any = this.pathobj.dir.split(path.sep);

        return {
            "workspaceFolder": vscode.workspace.rootPath,
            "workspaceFolderBasename": vscode.workspace.name,
            "file": this.editor.document.fileName,
            "relativeFile": path.join(rfd[rfd.length - 1], this.pathobj.base),
            "relativeFileDirname": rfd[rfd.length - 1],
            "fileBasename": this.pathobj.base,
            "fileBasenameNoExtension": this.pathobj.name,
            "fileDirname": this.pathobj.dir,
            "fileExtname": this.pathobj.ext,
            "cwd": vscode.workspace.rootPath,
        };    
    }

    private async _insert()
    {
        let lineCount: number = this.editor.document.lineCount;
        let tmplpath: string | undefined = this.getTemplatePath("header");
        let tmplpath_body: string | undefined = this.getTemplatePath("body");
        let ret: string = "";

        if (tmplpath)
        {
            const text = await vscode.workspace.fs.readFile(vscode.Uri.file(tmplpath));

            if (text.toString().trim() !== "") {
                ret = art_template.render(text.toString(), Object.assign(
                    {
                        author: this.config.author,
                        create_time: this.getFileBirthDateTime(),
                        last_modified_by: this.config.author,
                        last_modified_time: this.getDateTime(),
                        template: this.getDefaultTemplatePath(),
                    },
                    this.config.other_config,
                    this.predefinedVariables(),
                ));
            }

            if (lineCount <= 1 && tmplpath_body) {
                const bodytext = await vscode.workspace.fs.readFile(vscode.Uri.file(tmplpath_body));

                if (bodytext.toString().trim() !== "") {
                    ret += art_template.render(bodytext.toString(), Object.assign(
                        {
                            template: this.getDefaultTemplatePath(),
                        },
                        this.config.other_config,
                        this.predefinedVariables(),
                    ));    
                }
            }

            if (ret.trim() !== "") {
                this.editor.edit(editobj => {
                    editobj.insert(new vscode.Position(0, 0), ret);
                });
    
                this.editor.document.save();       
            }
        }
        else
        {
            console.debug("Not found fileheader template: " + this.editor.document.fileName);
        }
    }

    insert(): void {
        if (!this.isHeaderExists()) {
            this.deleteEditorComments();
            this._insert();
            this.insertEndComments();                
        }
    }

    private _update(): void {
        let start: number = 0;
        let line: number = 0;

        this.editor.edit(editobj => {
           line = this.findStringLine("@Last Modified time:");

           if (line !== -1) {
                start = this.editor.document.lineAt(line).text.indexOf(":") + 1;
                editobj.replace(new vscode.Range(line, start, line, 100), " " + this.getDateTime());				
            }

            line = this.findStringLine("@Last Modified by:");

            if (line !== -1) {
                start = this.editor.document.lineAt(line).text.indexOf(":") + 1;
                editobj.replace(new vscode.Range(line, start, line, 100), "   " + this.config.author);				
            }

            if (vscode.version < "1.43.0") {
                this.editor.document.save();	
            }        
        });
    }

    update(): void {
        if (!this.isIgnore(this.config.ignore)) {
            if (this.isHeaderExists()) {
                if (this.editor.document.isDirty) {
                    this._update();
                }
            }
            else if (this.config.save) {
                this.insert();
            }
        }
    }

    update2(): void {
        if (!this.isIgnore(this.config.ignore)) {
            if (this.isHeaderExists()) {
                if (this.editor.document.isDirty) {
                    this._update();
                }
            }
            else if (this.config.open) {
                this.insert();
            }
        }
    }
}

export { Template }

src/extension.ts
/**
 * @Author: JanKinCai
 * @Date:   2020-01-03 22:02:02
 * @Last Modified by:   JanKinCai
 * @Last Modified time: 2021-04-24 23:33:13
 */

// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import * as command from './command';

var template = require("art-template");


// Suffix ---> Template name
var file_suffix_mapping: any = {
    ".as": "ActionScript",
    ".scpt": "AppleScript",
    ".asp": "ASP",
    ".aspx": "ASP",
    ".bat": "Batch File",
    ".cmd": "Batch File",
    ".c": "C",
    ".cs": "C#",
    ".cpp": "C++",
    ".clj": "Clojure",
    ".css": "CSS",
    ".D": "D",
    ".dart": "Dart",
    ".erl": "Erlang",
    ".go": "Go",
    ".groovy": "Groovy",
    ".hs": "Haskell",
    ".htm": "HTML",
    ".html": "HTML",
    ".java": "Java",
    ".js": "JavaScript",
    ".tex": "LaTeX",
    ".lsp": "Lisp",
    ".lua": "Lua",
    ".md": "Markdown",
    ".mat": "Matlab",
    ".m": "Objective-C",
    ".ml": "OCaml",
    ".p": "Pascal",
    ".pl": "Perl",
    ".php": "PHP",
    ".py": "Python",
    ".r": "R",
    ".rs": "Rust",
    ".rst": "RestructuredText",
    ".rb": "Ruby",
    ".scala": "Scala",
    ".scss": "SCSS",
    ".sh": "ShellScript",
    ".sql": "SQL",
    ".tcl": "TCL",
    ".txt": "Text",
    ".ts": "TypeScript",
    ".vue": "Vue",
    ".xml": "XML",
    ".yml": "YAML",
    ".yaml": "YAML",

	".h": "H"
};


template.defaults.imports.upper = function(value: string)
{
	return value.toUpperCase();
};


template.defaults.imports.lower = function(value: string)
{
	return value.toLowerCase();
};


template.defaults.imports.replace = function(value: string, searchValue: string, replaceValue: string)
{
	return value.replace(searchValue, replaceValue)
};


/**
 * getConfig
 *
 * @return any
 */
function getConfig(): any {
	return vscode.workspace.getConfiguration("fileheader");
}


function sleep(ms: number){
    return new Promise(resolve=>{
        setTimeout(resolve,ms)
    })
}



export function activate(context: vscode.ExtensionContext) {
	console.log('Congratulations, your extension "vscodefileheader" is now active!');

	command.syncTemplateCommand(getConfig(), file_suffix_mapping);

	let disposable = vscode.commands.registerCommand('extension.fileheader', () => {
		command.updateTemplateCommand(getConfig, file_suffix_mapping);
	});
	context.subscriptions.push(disposable);

	disposable = vscode.commands.registerCommand('extension.synctemplate', () => {
		command.syncTemplateCommand(getConfig(), file_suffix_mapping);
	});
	context.subscriptions.push(disposable);

	disposable = vscode.commands.registerCommand('extension.createtemplate', () => {
		command.createTemplateCommand(getConfig(), file_suffix_mapping);
	});
	context.subscriptions.push(disposable);

	disposable = vscode.commands.registerCommand('extension.opentemplate', () => {
		command.openTemplateCommand(getConfig(), file_suffix_mapping);
	});
	context.subscriptions.push(disposable);

	// Save
	vscode.workspace.onWillSaveTextDocument(() => {
		command.updateTemplateCommand(getConfig(), file_suffix_mapping);
	});

	// Open
	vscode.workspace.onDidOpenTextDocument(() => {
		setTimeout(function() {
			command.updateTemplateCommand2(getConfig(), file_suffix_mapping);	
		}, 100);
	});
}

// this method is called when your extension is deactivated
export function deactivate() {}
